sklearn/inspection/_h_statistic.py

```diff
@@ -0,0 +1,325 @@
+"""Friedman and Popescu's H-Statistic"""
+
+# Authors: The scikit-learn developers
+# SPDX-License-Identifier: BSD-3-Clause
+
+import itertools
+
+import numpy as np
+from scipy import sparse
+
+from ..base import is_classifier, is_regressor
+from ..utils import Bunch, check_array
+from ..utils._indexing import _get_column_indices, _safe_assign, _safe_indexing
+from ..utils._param_validation import (
+    HasMethods,
+    Integral,
+    Interval,
+    Real,
+    validate_params,
+)
+from ..utils.random import sample_without_replacement
+from ..utils.validation import _check_sample_weight, check_is_fitted
+
+
+def _calculate_pd_brute_fast(
+    pred_fun, X, feature_indices, grid, sample_weight=None, reduce_binary=False
+):
+    """Fast version of _calculate_partial_dependence_brute()
+
+    Returns np.array of size (n_grid, output_dim).
+    """
+
+    # X is stacked n_grid times, and grid columns are replaced by replicated grid
+    n = X.shape[0]
+    n_grid = grid.shape[0]
+
+    X_stacked = _safe_indexing(X, np.tile(np.arange(n), n_grid), axis=0)
+    grid_stacked = _safe_indexing(grid, np.repeat(np.arange(n_grid), n), axis=0)
+
+    # TODO: remove via https://github.com/scikit-learn/scikit-learn/issues/28931
+    if hasattr(X, "iloc"):  # pandas<2 does not allow "values" to have repeated indices
+        grid_stacked = grid_stacked.reset_index(drop=True)
+        X_stacked = X_stacked.reset_index(drop=True)
+    _safe_assign(X_stacked, values=grid_stacked, column_indexer=feature_indices)
+
+    preds = pred_fun(X_stacked)
+
+    # Drop reduntant 0 class in binary classification
+    if reduce_binary:
+        preds = preds[:, 1]
+
+    # Partial dependences are averages per grid block
+    pd_values = np.average(
+        preds.reshape(n_grid, preds.shape[0] // n_grid, -1),
+        axis=1,
+        weights=sample_weight,
+    )
+
+    return pd_values
+
+
+def _calculate_pd_over_data(
+    pred_fun, X, feature_indices, sample_weight=None, reduce_binary=False
+):
+    """Calculates centered partial dependence over the data distribution.
+
+    It returns a numpy array of size (n, output_dim).
+    """
+
+    # Select grid columns and remove duplicates (will compensate below)
+    grid = _safe_indexing(X, feature_indices, axis=1)
```

Comment:
This is where I'm thinking that we could reuse the `_grid_from_X` instead. I don't know if taking quantile will actually have a statistical impact?